xcm/xcm-simulator/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Test kit to simulate cross-chain message passing and XCM execution
+
+pub use codec::Encode;
+pub use paste;
+
+pub use frame_support::{traits::Get, weights::Weight};
+pub use sp_io::TestExternalities;
+pub use sp_std::{cell::RefCell, marker::PhantomData};
+
+pub use polkadot_core_primitives::BlockNumber as RelayBlockNumber;
+pub use polkadot_parachain::primitives::{
+	DmpMessageHandler as DmpMessageHandlerT, Id as ParaId, XcmpMessageFormat,
+	XcmpMessageHandler as XcmpMessageHandlerT,
+};
+pub use polkadot_runtime_parachains::{
+	dmp,
+	ump::{self, MessageId, UmpSink, XcmSink},
+};
+pub use xcm::{v0::prelude::*, VersionedXcm};
+pub use xcm_executor::XcmExecutor;
+
+pub trait TestExt {
+	fn new_ext() -> sp_io::TestExternalities;
+	fn reset_ext();
+	fn execute_with<R>(execute: impl FnOnce() -> R) -> R;
+}
+
+pub enum MessageKind {
+	Ump,
+	Dmp,
+	Xcmp,
+}
+
+pub fn encode_xcm(message: Xcm<()>, message_kind: MessageKind) -> Vec<u8> {
+	match message_kind {
+		MessageKind::Ump | MessageKind::Dmp => VersionedXcm::<()>::from(message).encode(),
+		MessageKind::Xcmp => {
+			let fmt = XcmpMessageFormat::ConcatenatedVersionedXcm;
+			let mut outbound = fmt.encode();
+
+			let encoded = VersionedXcm::<()>::from(message).encode();
+			outbound.extend_from_slice(&encoded[..]);
+			outbound
+		},
+	}
+}
+
+#[macro_export]
+macro_rules! decl_test_relay_chain {
```

Comment:
It would be great if you could add some doc comments for these macros.

Reply:
lets follow up on that :)

https://github.com/paritytech/polkadot/issues/3542